### DISCUSSION TITLE
Parser fails in unexpected way

### DISCUSSION THREAD

**Discussion by purpl3F0x** [zesterer/chumsky#663] [Q&A] [answered]
Hello I have implemented a [parser](https://github.com/purpl3F0x/compilers/blob/main/alan/src/parser.rs), and i am getting the following error. 
I can't figure out why this happens, since for example, I can parse something like that successfully `if ( (n < lcv) & (lcv > n) )...`
This also passes `f ( n % lcv == 0) {...` but adding the parentheses, since to be problematic.

(I have tried adding labels to everything in cond/expr parsers, but it doesn't seem to solve the "something else" to something more indicative.)

```c
Error: Unexpected token
    ╭─[./examples/factors.alan:15:20]
    │
 15 │     if ( (n % lcv) == 0) {
    │                    ─┬  
    │                     ╰── found == expected something else
────╯
Error: Unexpected token
    ╭─[./examples/factors.alan:25:5]
    │
 25 │     }
    │     ┬  
    │     ╰── found } expected end of input
────╯
```

**Comment by zesterer** [comment]
It's hard to know what's going to without knowing much about the parser's design. Are you treating parenthesised expressions as having the same precedence as atoms (i.e: literals, variable names, etc.)?

**Reply by purpl3F0x** [reply]
Thanks for your reply. Parenthesized expressions take priority. 
With (descending) precedence
|Operator| Suffix (operands)| priority |
|:----|:----|:----|
|+ − ! | prefix(1)| left|
|* / %|  infix(2)| left|
|+ −  | infix(2)| left|
|== != > < <= >=| infix(2)| none|
|\& | infix(2)| left|
|\||  infix(2)| left|
 

The grammar I (try to) implement is as follows:

![image](https://github.com/user-attachments/assets/3ec94430-e79f-4eda-8df9-7fce142dc9ee)
![image](https://github.com/user-attachments/assets/4af5dc2e-1af8-4094-8071-952984b26acd)

So `"(" cond ")"` is has the second to most priority int he conditions.

**Comment by purpl3F0x** [comment]
I think the problem lies in the fact that the second "(" is parsed as `"(" cond ")"`, but  it fails in the way after reaching the `==`. 

Printing the context in the error message seems to indicate that,
```
    ╭─▶ main() : proc
    ┆ ┆   
 15 │ ├─▶     if ( (n % lcv) == 0) {
    │ │            ──────┬────┬        
    │ │                  ╰───────────── while parsing this condition
    │ │                       │        
    │ ╰──────────────────────────────── while parsing this function definition
    │                         │        
    │                         ╰──────── found == expected something else
    ```

**Reply by purpl3F0x** [reply]
I found what is probably causing the issue, after removing the error recovery from the `let atom =...` (line 160), it seems to parse as intended. I suppose the error recovery gets in the way of `(cond)`, rule falling back to the `(expr) op expr` rule.

I don't know if this intended behavior.

```rust
          .recover_with(via_parser(nested_delimiters(
               Token::ParenthesisOpen,
               Token::ParenthesisClose,
               [(Token::BraceOpen, Token::BraceClose), (Token::BracketOpen, Token::BracketClose)],
               |_| (ConditionAST { kind: ConditionKind::Error, span: Span::new(0, 0) }),
          )))
```

**Reply by zesterer** [reply, marked as answer]
It is intended behaviour, although admittedly not always desirable behaviour. Chumsky will prefer branches that result in fewer errors, but the exponential explosion of possible parse paths means that the parser may overly eagerly commit to a recovery pattern, missing an alternative valid parse. It's important to have error recovery occur as high up the stack as you reasonably can, and to not be overly general, so that the parser gets the opportunity to try all valid approaches to parsing the syntax before falling back to recovery.

**Reply by purpl3F0x** [reply]
I see. Thank you for the explanation.
